packages/rs-drive-abci/src/validation/state_transition/identity_create.rs

```diff
@@ -120,84 +141,72 @@ impl StateTransitionValidation for IdentityCreateTransition {
+        }
 
-        if asset_lock_already_found {
-            let outpoint = OutPoint::from(outpoint);
-            return Ok(ConsensusValidationResult::new_with_error(
-                ConsensusError::BasicError(
-                    BasicError::IdentityAssetLockTransactionOutPointAlreadyExistsError(
-                        IdentityAssetLockTransactionOutPointAlreadyExistsError::new(
-                            outpoint.txid,
-                            outpoint.vout as usize,
-                        ),
-                    ),
-                ),
-            ));
+        validation_result.merge(validate_asset_lock_proof_state(
+            &self.asset_lock_proof,
+            platform,
+            tx,
+        )?);
+
+        if !validation_result.is_valid() {
+            return Ok(validation_result);
         }
 
-        validation_result.add_errors(
-            validate_unique_identity_public_key_hashes_state(
-                self.public_keys.as_slice(),
-                drive,
-                tx,
-            )?
-            .errors,
-        );
+        validation_result.merge(validate_unique_identity_public_key_hashes_state(
+            self.public_keys.as_slice(),
+            drive,
+            tx,
+        )?);
+
+        if !validation_result.is_valid() {
+            return Ok(validation_result);
+        }
 
         // Now we should check the state of added keys to make sure there aren't any that already exist
-        validation_result.add_errors(
-            validate_unique_identity_public_key_hashes_state(
-                self.public_keys.as_slice(),
-                drive,
-                tx,
-            )?
-            .errors,
-        );
+        validation_result.merge(validate_unique_identity_public_key_hashes_state(
+            self.public_keys.as_slice(),
+            drive,
+            tx,
+        )?);
 
         if !validation_result.is_valid() {
             return Ok(validation_result);
         }
 
-        self.transform_into_action(platform, tx)
+        self.transform_into_action(platform, context, tx)
     }
 
     fn transform_into_action<C: CoreRPCLike>(
         &self,
         platform: &PlatformRef<C>,
+        context: &ValidationDataShareContext,
         _tx: TransactionArg,
     ) -> Result<ConsensusValidationResult<StateTransitionAction>, Error> {
         let mut validation_result = ConsensusValidationResult::<StateTransitionAction>::default();
 
-        let tx_out_validation = self
-            .asset_lock_proof
-            .fetch_asset_lock_transaction_output_sync(platform.core_rpc)?;
-        if !tx_out_validation.is_valid() {
+        let asset_lock_validation_result = context
+            .asset_lock_output
+            .as_ref()
+            .map(|tx_out| Ok(ConsensusValidationResult::new_with_data(tx_out.clone())))
+            .unwrap_or_else(|| {
+                fetch_asset_lock_transaction_output_sync(platform.core_rpc, &self.asset_lock_proof)
+            })?;
```

Comment:
There is a misunderstanding about what each section is supposed to do, in this case the transform into action is where you should have all network based processing and validation. Hence there is no need for context. Transform into action is state transition + state + core state -> state transition action.

Reply:
Do we want to fetch asset lock twice?

Reply:
No, only once.